src/bag.js

```diff
@@ -108,16 +109,91 @@ export default class Bag {
       return new ReadResult(topic, message, timestamp, data, chunkOffset, chunkInfos.length);
     }
 
-    for (let i = 0; i < chunkInfos.length; i++) {
-      const info = chunkInfos[i];
-      const messages = await this.reader.readChunkMessagesAsync(
-        info,
-        filteredConnections,
-        startTime,
-        endTime,
-        decompress
-      );
-      messages.forEach((msg) => callback(parseMsg(msg, i)));
+    // Process the list of chunks such that no more than a maximum amount of bytes are being used and at
+    // least one chunk is processed at a time
+    function* processChunks(chunkInfos, processedCb) {
```

Comment:
I'm thinking a generator may be making this more complicated than it needs to be so I'll probably change it before this is considered complete.